src/bpmncwpverify/core/state.py

```diff
@@ -96,212 +98,403 @@ def _parse_state(parser: StateParser) -> Result[StateParser.StateContext, Error]
         return Failure(failure_value)
 
 
+class DeclLoc:
+    __slots__ = ["line", "col"]
+
+    def __init__(self, line: Maybe[int], col: Maybe[int]) -> None:
+        self.line = line
+        self.col = col
+
+
+class AllowedValueDecl(DeclLoc):
+    __slots__ = ["value"]
+
+    def __init__(
+        self, value: str, line: Maybe[int] = Nothing, col: Maybe[int] = Nothing
+    ) -> None:
+        super().__init__(line, col)
+        self.value = value
+
+
+class ConstDecl(DeclLoc):
+    __slots__ = ["id", "type_", "init"]
+
+    def __init__(
+        self,
+        id: str,
+        type_: str,
+        init: AllowedValueDecl,
+        line: Maybe[int] = Nothing,
+        col: Maybe[int] = Nothing,
+    ) -> None:
+        super().__init__(line, col)
+        self.id = id
+        self.type_ = type_
+        self.init = init
+
+
+class EnumDecl(DeclLoc):
+    __slots__ = ["id", "values"]
+
+    def __init__(
+        self,
+        id: str,
+        values: list[AllowedValueDecl],
+        line: Maybe[int] = Nothing,
+        col: Maybe[int] = Nothing,
+    ) -> None:
+        super().__init__(line, col)
+        self.id = id
+        self.values = values
+
+
+class VarDecl(DeclLoc):
+    __slots__ = ["id", "type_", "init", "values", "line", "col"]
```

Comment:
This is awesome! So, we are forcing the user to pass in these wrappers into the builder `with_` methods? I think this is a cool approach